File: longest-univalue-path.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestUnivaluePath(self, root: TreeNode) -> int:
        if root is None:
            return 0
        
        path = 0
        
        def dfs(node=root):
            if node.left is None and node.right is None:
                return 0
            
            left = dfs(node.left) if node.left else 0
            right = dfs(node.right) if node.right else 0
            
            left = left + 1 if node.left and node.val == node.left.val else 0
            right = right + 1 if node.right and node.val == node.right.val else 0
                
            nonlocal path
            path = max(path, left + right)
            return max(left, right)
        
        dfs()
        return path

File: longest-univalue-path.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int longest = 0;

    public int longestUnivaluePath(TreeNode root) {
        dfs(root);
        return longest;
    }

    public int dfs(TreeNode node) {
        if(node == null)
            return 0;
        
        int left = dfs(node.left);
        int right = dfs(node.right);

        if(node.left != null && node.left.val == node.val)
            left++;
        else
            left = 0;

        if(node.right != null && node.right.val == node.val)
            right++;
        else
            right = 0;
        
        longest = Math.max(longest, left + right);
        return Math.max(left, right);
    }
}